examples/abc/provider_b_hub/tests/provider/test_provider_b_provider.py

```diff
@@ -0,0 +1,47 @@
+"""pact test for user service client"""
+import logging
+
+import pytest
+
+from pact import Verifier
+
+log = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+# For the purposes of this example, the broker is started up as a fixture defined
+# in conftest.py. For normal usage this would be self-hosted or using Pactflow.
+PACT_BROKER_URL = "http://localhost"
+PACT_BROKER_USERNAME = "pactbroker"
+PACT_BROKER_PASSWORD = "pactbroker"
+
+# For the purposes of this example, the FastAPI provider will be started up as
+# a fixture in conftest.py ("server"). Alternatives could be, for example
+# running a Docker container with a database of test data configured.
+# This is the "real" provider to verify against.
+PROVIDER_HOST = "127.0.0.1"
+PROVIDER_PORT = 8000
+PROVIDER_URL = f"http://{PROVIDER_HOST}:{PROVIDER_PORT}"
+PROVIDER_NAME = "ProviderBHub"
+
+
+@pytest.fixture
+def broker_opts():
+    return {
+        "broker_username": PACT_BROKER_USERNAME,
+        "broker_password": PACT_BROKER_PASSWORD,
+        "broker_url": PACT_BROKER_URL,
+        "publish_version": "3",
+        "publish_verification_results": True,
+    }
+
+
+def test_provider_b_hub_against_broker(server, broker_opts, **kwargs):
+    verifier = Verifier(provider=PROVIDER_NAME, provider_base_url=PROVIDER_URL)
+
+    success, logs = verifier.verify_with_broker(
+        **broker_opts,
+        verbose=True,
+        provider_states_setup_url=f"{PROVIDER_URL}/_pact/provider_states",
```

Comment:
I'm trying to understand `provider_states_setup_url`, can you explain what it does or point me to documentation about it?

Reply:
I didn't manage to find as good documentation as I would like for that when I was trying to get it working, descriptions and examples in some languages but it probably needs more work. https://docs.pact.io/implementation_guides/python/readme#provider-states

The TLDR is you create an extra endpoint (created here: examples/abc/provider_b_hub/tests/pact_provider_b.py). When the verifier performs the tests, for each test it FIRST does a POST to that endpoint with {"state": "some books exist"} etc. You then define whatever you need to there to get the service into an appropriate state, so when the ACTUAL request is made, it can behave as desired.

I like how this is setup with FastAPI, to be able to only have this extra route for the purposes of testing, it gives a nice way to not have some super admin drop all tables and setup endpoint accidentally left open!

Reply:
Revisiting this, I think I can do something similar in Django. This is intriguing. Thanks!